src/metadataGeneration/typeResolver.ts

```diff
@@ -185,40 +191,56 @@ export class TypeResolver {
       return literalType;
     }
 
-    let referenceType: Tsoa.ReferenceType;
     if (typeReference.typeArguments && typeReference.typeArguments.length > 0) {
       this.typeArgumentsToContext(typeReference, typeReference.typeName, this.context);
     }
 
-    referenceType = this.getReferenceType(typeReference.typeName as ts.EntityName, this.extractEnum, typeReference.typeArguments);
```

Comment:
I don't really see why this has any benefits over the current implementation and it does not warrant all this new logic to avoid type casts in my opinion.
Honestly, I am already struggling with the scope of the type aliases. This feels like an additional punishment.

Reply:
The only advantage it provides is that refObjects definitely don’t have an `enums` property and refEnums definitely don’t have a `properties` property.

In the current code, the interface has all properties as nullable.

But yes, I hear you. The ramifications of the tagged union approach were much more than I expected. So if you don’t see an advantage in the approach, then we can close this PR.

Reply:
@dgreene1 I think I didn't clearly state what I meant by "this".

I am specifically referring to the change of logic below the line I commented on.
I approve of the type changes (minus the naming). I would've liked to see a narrower scope basically.

After checking it out locally now, I think I am more ok with the logic here, maybe I am too biased because of all the rebasing I'll have to do, keep that in mind.